### ISSUE TITLE
no-unused-variable rule should not be deprecated

### ISSUE THREAD

**Issue by UselessPickles** [palantir/tslint#4100]
### Bug Report

- __TSLint version__: 5.11.0
- __TypeScript version__: 2.9.2
- __Running TSLint via__: CLI

#### Actual behavior

"no-unused-variable is deprecated. Since TypeScript 2.9. Please use the built-in compiler checks instead."

#### Expected behavior

This rule should not be deprecated.

The TypeScript compiler check is not a replacement for a linter. The lint rule has benefits that cannot be provided by the compiler option:
* Linter supports "fix" behavior.
* Linter supports enabling/disabling the rule with file, block, or line granularity, while the compiler option applies to the entire project.

Here's a comment from a TypeScript developer that is applicable: https://github.com/Microsoft/TypeScript/issues/11051#issuecomment-248971532

**Comment by donaldpipowitch** [comment]
Just for discoverability here the link to the pull request where this was deprecated: https://github.com/palantir/tslint/pull/3919

> Linter supports "fix" behavior.

I _think_ TypeScript should support this now as well in the form of "suggestions", but I tried to use them and failed (https://github.com/Microsoft/TypeScript/issues/13408#issuecomment-412438318). It would be nice to see a migration guide from the TSLint team which covers this, if they still want to deprecate this rule. (cc @JKillian )

> Linter supports enabling/disabling the rule with file, block, or line granularity, while the compiler option applies to the entire project.

Valid point 👍

**Comment by DannyDelott** [comment]
Another use-case for un-deprecating this..

I have a function that is called with 3 arguments, I need to declare the second one to get access to the third:

```js
const fn = (first, unusedSecond, third) => {
  return `${first} ${third}`;
};
```

This occurs often in the React ecosystem, where side-effect middleware like `redux-thunk` and `redux-observable` can be configured with extra arguments:

```js
const actionCreator = (dispatch, unusedGetState, extraArg) => { ... };
```

With the lint rule, we can specify a pattern matcher for ignoring unused variables:

```
"no-unused-variable": [
  true,
  { "ignore-pattern": "^unused" }
]
```

I am not aware of an approach that allows this using just the Typescript compiler.

**Comment by UselessPickles** [comment]
Aside from the specific use case examples of how the compiler option cannot fully replace the linter rule, I really want to make sure that the opinion of the TypeScript team on matters like this is considered. Here's the relevant comment link again: https://github.com/Microsoft/TypeScript/issues/11051#issuecomment-248971532

Basically, some of the compiler options are things they made available just because it was so easy to do; not because it's really the compiler's job to check these things. They will never expand functionality of any of these compiler checks to support things like conditionally enabling/disabling a rule based on comments, configuration of patterns, etc.

**Comment by aindeev** [comment]
The "fix" behaviour has become fundamental in our workflow. Enabling TS "noUnusedLocals" means we have to manually fix everything. 

Please consider un-deprecating this rule

**Comment by karfau** [comment]
Another aspect is that the compiler option `noUnusedLocals` can be disabled in `tsconfig.json` that is used for running tslint. So maybe there is a way to check for this flag and only log a deprecation warning when it is enabled? (Which will of course not be true for older versions of typescript that tslint still supports looking at it's peer dependencies.)

**Comment by chris-allnutt** [comment]
Came across this today as well.  Would like to see this un-deprecated.  I prefer to catch these issues before I compile.

**Comment by ajafff** [comment]
@iensu 

>  If it's doable in eslint, shouldn't it be doable for tslint as well?

It is. I did this a while back: https://github.com/ajafff/tslint-consistent-codestyle/blob/master/docs/no-unused.md
Note that it won't check `private` properties and methods.

In older versions of TypeScript (<2.9) it was really difficult to correctly detect unused imports as they may be implicitly required by TypeScript's declaration emitter. That's why the rule has an option to ignore imports completely. This is no longer necessary in more recent versions of TypeScript.

**Comment by iensu** [comment]
@ajafff that worked like a charm, I'm going to addyour package to our dev deps. Thanks a lot!

**Comment by cyberhck** [comment]
I have similar but one more usecase, I'm linting my project on pre-commit hook, and testing and building on pre-push hook, I'd like to catch this one and fix it even before I commit, I don't want to see any additional commit, (unless user does a soft reset), but still, it's a waste of time, I can't do typescript build on commit hook because project is relatively large and it'll take time to compile :)

**Comment by JoshuaKGoldberg** [comment]
Another fun, contentious issue 😊.

The `no-unused-variable` rule as it is now is very difficult to maintain and can wreak havoc on other tools, as already noted in this thread. If TSLint core is to have a version of it, it should either...
* ...be something like ajafff's `tslint-consistent-codestyle` that does its own checking _(what about privates?)_ 
* ...still mirror TypeScript's diagnostics but not wreak havoc on other tools _(how?)_

**Comment by donaldpipowitch** [comment]
If we would have access to the language service we could just _use_ TypeScripts diagnostics instead of mirroring them, if I'm not mistaken? I don't know why, but it seems like access to the language service was removed in v5 (I once created an issue for this: https://github.com/palantir/tslint/issues/4232).

**Comment by glen-84** [comment]
@DannyDelott,

You can use an underscore, and TypeScript will not complain:

```ts
const fn = (first, _second, third) => {
  return `${first} ${third}`;
};
```

@dolanmiu, (in reply to https://github.com/palantir/tslint/pull/3919#issuecomment-425995016)

You can use `@ts-ignore`:

```ts
// @ts-ignore
@HostBinding('class') private classes = 'my-theme';
```

Note: This will suppress *all* TypeScript errors on that line. See TypeScript issue: [Support @ts-ignore for specific errors](https://github.com/Microsoft/TypeScript/issues/19139).

**Comment by sahilanguralla** [comment]
+1 for the issue.

In my scenario, I want to prevent committing of code with any lint/typescript errors. I am using `lint-staged` and doing simple lint check on all files. Since it is a typescript compile-time error, it is not thrown and the commit happens with erroneous code in it.

As a workaround, I am running compile command without emitting files `tsc --noEmit -p`. This throws compile-time errors, if any, in development phase itself. But, I believe this should be part of linting phase itself like as in case of eslint.

**Comment by Serginho** [comment]
+1 me too

Example: I'm developing some feature. Comment line, watchers recompile and I see warnings.
I prefer this in the lint process.

**Comment by dmccown1500** [comment]
Also the fix functionality is CLUTCH! Your guys have an awesome library, BTW!

**Comment by lukewlms** [comment]
[tslint-etc](https://github.com/cartant/tslint-etc) handles this use case admirably.

**Comment by lukewlms** [comment]
[tslint-etc](https://github.com/cartant/tslint-etc) handles this use case [admirably for us](https://medium.com/2-minute-revolution-developer-blog/our-full-linting-and-commit-setup-for-typescript-prettier-tslint-lint-staged-husky-e442ad3f4be).

The author is also [very](https://github.com/cartant/tslint-etc/issues/14) [responsive](https://github.com/cartant/tslint-etc/issues/13) to [issues](https://github.com/cartant/tslint-etc/issues/15).

**Comment by fibo** [comment]
In my opinion no-unused-variable should be included in tslint recommended rules too, the fact that it is deprecated, then undeprecated, then now in v5.17 it says again it is deprecated, it is one of the reasons why I am considering take a look to another linter for TS.

At every developer out there: think twice about retrocompatibility, please **do not** remove things!

> May [The Slack](https://www.youtube.com/watch?v=KE5FVKbQ4ow) be with you

**Comment by JoshuaKGoldberg** [comment]
Note: per #4534, this issue will be closed in less than a month if no PR is sent to add the rule. If you _really_ need the rule, custom rules are always an option and can be maintained outside this repo!

**Comment by fibo** [comment]
Other than my pleaseCareAboutRetroCompatibilityRant tslint is a great piece of software and a must have in all my TypeScript projects.

For those who want a similar feature you can add this to your tsconfig.json

    "noUnusedLocals": true,
    "noUnusedParameters": true,

See, for instance, this tsconfig I am using: https://github.com/go-seven/go-seven.com/blob/master/tsconfig.json

I think it should be enabled in every TypeScript project even from the beginning prototyping phase.

**Comment by cyberhck** [comment]
@JoshuaKGoldberg thanks for reminder, sent a PR hopefully it's enough.

@fibo 
I feel like you didn't read the whole thing, we want this because we want to test before we do a build, most of the people use precommit hooks for linting and prepush for build and test, if you can detect a problem before you commit, it's always better, you can still have your build detect that if that works better for you. On your project, wouldn't you be more glad if you could detect that before commit?

**Comment by JoshuaKGoldberg** [comment]
@cyberhck to be clear, what we're accepting are PRs to resolve the technical limitations that originally caused the rule to be deprecated in the first place.

tl;dr:
* *Everybody wants the rule to not be deprecated*, maintainers included
* *There are technical blockers that need to be overcome to un-deprecate the rule*
* You can always fork TSLint and remove the deprecation warning in your fork

It's a little heartbreaking seeing everyone post how nice it would be to be able to use the rule. There's very little that can be done until the above blockers are either fixed or found to no longer be relevant. Per #4534 you should divert your energy towards the [typescript-eslint](http://typescript-eslint.io) project.

I'm going to mark supporting comments as off topic so folks are more likely to read this one.

**Comment by dakotahawkins** [comment]
@JoshuaKGoldberg Thanks for the tl;dr! Would it be possible to summarize/link to the technical blockers, and add them to your explanation?

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is being deprecated** and no longer accepting pull requests for new rules. See #4534. 😱

If you'd like to see this rule implemented, you have two choices:
* **Recommended:** Use [typescript-eslint](https://typescript-eslint.io) to lint your TypeScript code with ESLint ✅
* **Legacy:** Write it as a [custom TSLint rule](https://palantir.github.io/tslint/develop/custom-rules/) 🤷‍♂️

👋 It was a pleasure open sourcing with you!

_If you believe this message was posted here in error, please comment so we can re-open the issue!_

**Comment by JoshuaKGoldberg** [comment]
Linking here: https://github.com/palantir/tslint/issues/3671 looks like good supporting docs for why it was deprecated.

**Comment by libinvarghese** [comment]
You could use [`tslint-consistent-codestyle` `no-unused`](https://github.com/ajafff/tslint-consistent-codestyle/blob/master/docs/no-unused.md) rule instead

**Comment by beenotung** [comment]
@libinvarghese extends `tslint-consistent-codestyle` with rule `no-unused` seems cannot be auto fixed.
I'm extending `tslint-etc` with rule `no-unused-declaration` instead.

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
